Index: libmspack/mspack/chmd.c
===================================================================
--- libmspack.orig/mspack/chmd.c
+++ libmspack/mspack/chmd.c
@@ -746,7 +746,7 @@ static int search_chunk(struct mschmd_he
 	    /* compare filename with entry QR points to */
 	    p = &chunk[entries_off + (M ? EndGetI16(start - (M << 1)) : 0)];
 	    READ_ENCINT(name_len);
-	    if (p + name_len > end) goto chunk_end;
+	    if (name_len > end - p) goto chunk_end;
 	    cmp = compare(filename, (char *)p, fname_len, name_len);
 
 	    if (cmp == 0) break;
@@ -783,7 +783,7 @@ static int search_chunk(struct mschmd_he
     *result = NULL;
     while (num_entries-- > 0) {
 	READ_ENCINT(name_len);
-	if (p + name_len > end) goto chunk_end;
+	if (name_len > end - p) goto chunk_end;
 	cmp = compare(filename, (char *)p, fname_len, name_len);
 	p += name_len;
 
